### PR TITLE
Use patch instead of replace for prowjobs in plank

### LINKED ISSUE #13355: Plank tries to update a ProwJob multiple times in the same sync loop.
prow.k8s.io is seeing errors like this:
```
errors syncing: [body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"Operation cannot be fulfilled on prowjobs.prow.k8s.io \"6eaca2d6-a1be-11e9-83df-c67ad48cda3d\": the object has been modified; please apply your changes to the latest version and try again","reason":"Conflict","details":{"name":"6eaca2d6-a1be-11e9-83df-c67ad48cda3d","group":"prow.k8s.io","kind":"prowjobs"},"code":409}
 body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"Operation cannot be fulfilled on prowjobs.prow.k8s.io \"6eadbc5e-a1be-11e9-83df-c67ad48cda3d\": the object has been modified; please apply your changes to the latest version and try again","reason":"Conflict","details":{"name":"6eadbc5e-a1be-11e9-83df-c67ad48cda3d","group":"prow.k8s.io","kind":"prowjobs"},"code":409}
 body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"Operation cannot be fulfilled on prowjobs.prow.k8s.io \"6eae64d4-a1be-11e9-83df-c67ad48cda3d\": the object has been modified; please apply your changes to the latest version and try again","reason":"Conflict","details":{"name":"6eae64d4-a1be-11e9-83df-c67ad48cda3d","group":"prow.k8s.io","kind":"prowjobs"},"code":409}
], errors reporting: []
```

I think this suggests that we are trying to update a ProwJob multiple times during the same sync loop which we shouldn't be doing.

cc @stevekuznetsov @fejta

### PR DESCRIPTION
**Pull request by krzyzacy**
Since now plank is not the only source of truth can update a prowjob anymore. (crier, and maybe other controllers, in the future)

But the fields plank && crier can edit are mutually exclusive, so patch should fix #13355

/assign @cjwagner @stevekuznetsov

### CODE REVIEW

File: prow/plank/controller.go [hunk 1]
```diff
@@ -554,3 +557,29 @@ func getPodBuildID(pod *coreapi.Pod) string {
 	logrus.Warningf("BUILD_ID was not found in pod %q: streaming logs from deck will not work", pod.ObjectMeta.Name)
 	return ""
 }
+
+func (c *Controller) patchProwjob(srcPJ prowapi.ProwJob, destPJ prowapi.ProwJob) error {
+	srcPJData, err := json.Marshal(srcPJ)
+	if err != nil {
+		return fmt.Errorf("error marshal srcPJ: %v", err)
```
Comment (stevekuznetsov) [inline prow/plank/controller.go]: These errors do not expose enough about what is going on here

File: prow/plank/controller.go [hunk 2]
```diff
@@ -554,3 +557,29 @@ func getPodBuildID(pod *coreapi.Pod) string {
...
+		return fmt.Errorf("error marshal destPJ: %v", err)
+	}
+
+	patch, err := jsonpatch.CreateMergePatch(srcPJData, destPJData)
+	if err != nil {
+		return fmt.Errorf("error CreateMergePatch: %v", err)
+	}
+
+	if len(patch) == 0 {
+		logrus.Warnf("Empty merge patch: srcPJData: %s, destPJData: %s", string(srcPJData), string(destPJData))
```
Comment (stevekuznetsov) [inline prow/plank/controller.go]: This log line is going to be huge

**Comment by stevekuznetsov** [comment]
I also worry this still doesn't solve the issue if `crier` is updating status -- the patch we generate could just undo updates from other controllers if our last known state is old enough.

**Comment by krzyzacy** [comment]
> I also worry this still doesn't solve the issue if crier is updating status -- the patch we generate could just undo updates from other controllers if our last known state is old enough.

Plank and Crier updates different prowjob status fields, so patches from each other won't undo other controller's patch

> All we ever do is update status so a better approach would be to actually use that feature in the APIServer 

Migrate off our k8s client? Sure...

**Comment by stevekuznetsov** [comment]
But you aren't building a patch for those fields, you are building a diff patch for the entire object

**Comment by krzyzacy** [comment]
> But you aren't building a patch for those fields, you are building a diff patch for the entire object

Plank only changes these fields.

Say we have 

```
Prowjob
- plank: foo
- crier: foo
```

Plank patches it with `plank: bar` and crier patches it with `crier: bar` at the same time, it should still work. (In plank we basically get a prowjob, set a status field, and replace. When create the patch it won't give us any field that outside of plank)

**Comment by stevekuznetsov** [comment]
The race is:

1. plank reads current state 
2. crier patches state
3. plank calculates diff to make a patch from #1 to the dest state

**Comment by krzyzacy** [comment]
hummm, I don't think so since they are touch different status fields:

Plank reads
```
Prowjob
- plank: foo
- crier: foo
```

After crier reads/patches with patch `crier: bar`, upstream prowjob becomes
```
Prowjob
- plank: foo
- crier: bar
```

Local prowjob modification in plank:
```
Prowjob
- plank: foo -> bar
- crier: foo
```

Diff the prowjob plank reads and the one after the modification, we will got a patch like `plank: bar`, after we submit the patch to apiserver, we will get

```
Prowjob
- plank: bar
- crier: bar
```

Which is desired.

File: prow/plank/controller.go (hunk 1 above)
Comment (krzyzacy) [reply prow/plank/controller.go]: do you want to append the entire prowjob here?

Comment (stevekuznetsov) [reply prow/plank/controller.go]: No, that's not useful either. Something like "could not create JSON patch:". No admin knows what `srcPJ` is

File: prow/plank/controller.go (hunk 2 above)
Comment (krzyzacy) [reply prow/plank/controller.go]: it should not happen, removed

File: prow/plank/controller.go (hunk 1 above)
Comment (krzyzacy) [reply prow/plank/controller.go]: done

**Comment by krzyzacy** [comment]
PTAL

also cc @michelle192837

**Pull request merged**